bgpq3.8
.\" Copyright (c) 2007-2025 Alexandre Snarskii
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd Oct 27, 2008
.Dt BGPQ3 8
.Os
.Sh NAME
.Nm bgpq3
.Nd "bgp filtering automation tool"
.Sh SYNOPSIS
.Nm
.Op Fl h Ar host[:port]
.Op Fl I Ar laddr
.Op Fl S Ar sources
.Op Fl EPz
.Oo
.Fl f Ar asn |
.Fl F Ar fmt |
.Fl G Ar asn 
.Fl t
.Oc
.Op Fl 2346ABbDdJjHNnsXU
.Op Fl a Ar asn
.Op Fl r Ar len
.Op Fl R Ar len
.Op Fl m Ar max
.Op Fl W Ar len
.Ar OBJECTS
.Op "..."
.Op EXCEPT OBJECTS
.Sh DESCRIPTION
The
.Nm 
utility used to generate configurations (prefix-lists, extended
access-lists, policy-statement terms and as-path lists) based on RADB data.
.Pp
The options are as follows: 
.Bl -tag -width Ds
.It Fl 2
accept routes registered for as23456 (transition-as) (default: false)
.It Fl 3
assume that your device is asn32-safe.
.It Fl 4 
generate IPv4 prefix/access-lists (default).
.It Fl 6 
generate IPv6 prefix/access-lists (IPv4 by default).
.It Fl A
try to aggregate prefix-lists as much as possible (not all output
formats supported).
.It Fl a Ar asn
specify what asn shall be denied in case of empty prefix-list (OpenBGPD)
or included in empty as-list (JunOS).
.It Fl B
generate output in OpenBGPD format (default: Cisco)
.It Fl b
generate output in BIRD format (default: Cisco).
.It Fl d
enable some debugging output.
.It Fl D
use asdot notation for Cisco as-path access-lists.
.It Fl E
generate extended access-list (Cisco), policy-statement term using
route-filters (Juniper), [ip|ipv6]-prefix-list (Nokia) or prefix-sets
(OpenBGPd).
.It Fl f Ar number
generate input as-path access-list (use 0 to not enforce first AS)
.It Fl F Ar fmt
generate output in user-defined format.
.It Fl G Ar number
generate output as-path access-list.
.It Fl H
hyperaggregation (supernets-only) mode.
.It Fl h Ar host[:port]
host running IRRD database (default: whois.radb.net).
.It Fl I Ar laddr
source address used to connect to servers
.It Fl J
generate config for Juniper (default: Cisco).
.It Fl j
generate output in JSON format (default: Cisco).
.It Fl l Ar name 
name of generated entry.
.It Fl L Ar limit
limit recursion depth when expanding as-sets.
.It Fl m Ar len
maximum prefix-length of accepted prefixes (default: 32 for IPv4 and 
128 for IPv6).
.It Fl M Ar match 
extra match conditions for Juniper route-filters.
.It Fl n
generate config for Nokia SR OS MD-CLI (Cisco IOS by default)
.It Fl N
generate config for Nokia SR OS classic CLI (Cisco IOS by default).
.It Fl p
accept routes registered for private ASNs (default: disabled)
.It Fl P
generate prefix-list (default, backward compatibility).
.It Fl r Ar len
allow more specific routes starting with specified masklen too. 
.It Fl R Ar len
allow more specific routes up to specified masklen too. 
.It Fl s
generate sequence numbers in IOS-style prefix-lists.
.It Fl S Ar sources
use specified sources only (recommended: RADB,RIPE,APNIC).
.It Fl t
generate as-sets for OpenBGPD (OpenBSD 6.4+), BIRD and JSON formats,
as-list-groups for Juniper (JunOS 21+).
.It Fl T
disable pipelining. 
.It Fl W Ar len
generate as-path strings of no more than len items (use 0 for inifinity).
.It Fl U
generate config for Huawei devices (Cisco IOS by default)
.It Fl X
generate config for Cisco IOS XR devices (plain IOS by default).
.It Fl z
generate route-filter-lists (JunOS 16.2+).
.It Ar OBJECTS 
means networks (in prefix format), autonomous systems, as-sets and route-sets.
.It Ar EXCEPT OBJECTS

CHANGES
untagged yet
    - implement -I laddr: bind to local address when connecting.
    Mostly based on pull by Jochen Kornitzky with minor changes.
    - allow asn validation (-w) work for as-set/as-list output too.

0.1.38 (2025-05-11):
    - cosmetic bugfix: missng -z documentation in usage() and exclusive().
    - handle rare cases with leading/middle spaces in IRRd answers.
    - fail early when request can't fit buffer.
    - implement as-lists for JunOS (-t together with -J).
    Note: empty as-lists are automatically removed, so it may make sense
    to specify "replacement asn" with option -a <asn> in this mode.
    - cleanup note about LACNIC in README

0.1.37 (2024-04-01), for some reason never published
    - fix missing reference to config.h in sx_slentry.c
    - initial support for mixed address-family (-x) feature: this allows
    you to generate prefix-lists/route-filters/.. for IPv4 and IPv6 together.
    Note: this feature is currently limited to Juniper/JSON/User-Defined
    output formats and is not compatible with maximum prefix length (-m)
    and more-specific (-r/-R) features.

0.1.36.1 (2021-09-27):
    - minor bugfix: update version number in configure and bgpq3.spec.

0.1.36 (2021-09-22):
    - initial support for as-transparent route-servers in import as-path
    generation: now you can use -f 0 to not enforce first asn in path.
    No actual changes for Bird and JSON, not yet supported for Nokia
    (including MD) and OpenBGPD. Use case reported by Mikhail Grishin.
    - correct sequencing in empty prefix-lists. Reported by Brooks Swinnerton.
    - fix empty route-filter-lists generation. Reported by Antti Ristimäki.
    - minor documentation cleanup: bgpq3 supports much more vendors
    than just Cisco and Juniper. Copyright years updated.
    - Nokia SR OS "classic" actually supports aggregation and
    and more-specific filtering in prefix-lists. Thanks to mfisher128
    for reporting.
    - change log level for prefixes with wrong address family from error
    to debugging: it's perfectly correct to have prefixes of different
    families in route-sets. Thanks to Jay Ford for suggestion.
    - increase select timeout to 30 sec.
    - correct spacing for IOS-XR. Reported by Mathias Wolkert.
    - hyperaggregation (supernets-only) mode.

bgpq3.c
#if HAVE_CONFIG_H
#include <config.h>
#endif

#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <ctype.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <unistd.h>

#include "bgpq3.h"
#include "sx_report.h"

extern int debug_expander;
extern int debug_aggregation;
extern int pipelining;
extern int expand_as23456;
extern int expand_special_asn;

int
usage(int ecode)
{
	printf("\nUsage: bgpq3 [-h host[:port]] [-S sources] [-P|E|G <num>|f <num>|t]"
		" [-2346ABbDdHJjNnwXxz] [-R len] <OBJECTS>...\n");
	printf(" -2        : allow routes belonging to as23456 (transition-as) "
		"(default: false)\n");
	printf(" -3        : assume that your device is asn32-safe\n");
	printf(" -4        : generate IPv4 prefix-lists (default)\n");
	printf(" -6        : generate IPv6 prefix-lists (IPv4 by default)\n");
	printf(" -A        : try to aggregate prefix-lists/route-filters\n");
	printf(" -a asn    : replacement asn for empty as-lists(JunOS)/prefix-lists(OpenBSD)\n");
	printf(" -B        : generate OpenBGPD output (Cisco IOS by default)\n");
	printf(" -b        : generate BIRD output (Cisco IOS by default)\n");
	printf(" -D        : use asdot notation in as-path (Cisco only)\n");
	printf(" -d        : generate some debugging output\n");
	printf(" -E        : generate extended access-list(Cisco), "
		"route-filter(Juniper)\n"
		"             [ip|ipv6]-prefix-list (Nokia) or prefix-set (OpenBGPD)"
		"\n");
	printf(" -F fmt    : generate output in user-defined format\n");
	printf(" -f number : generate input as-path access-list (use 0 to not "
		"enforce first AS)\n");
	printf(" -G number : generate output as-path access-list\n");
	printf(" -H        : hyper-aggregation (supernets only) mode\n");
	printf(" -h host   : host running IRRD software (whois.radb.net by "
		"default)\n"
		"             (use host:port to specify alternate port)\n");
	printf(" -I laddr  : bind to local address\n");
	printf(" -J        : generate config for JunOS (Cisco IOS by default)\n");
	printf(" -j        : generate JSON output (Cisco IOS by default)\n");
	printf(" -M match  : extra match conditions for JunOS route-filters\n");
	printf(" -m len    : maximum prefix length (default: 32 for IPv4, "
		"128 for IPv6)\n");
	printf(" -L depth  : limit recursion depth (default: unlimited)\n"),
	printf(" -l name   : use specified name for generated access/prefix/.."
		" list\n");
	printf(" -N        : generate config for Nokia SR OS classic CLI "
		"(Cisco IOS by default)\n");
	printf(" -n        : generate config for Nokia SR OS MD-CLI (Cisco IOS "
		"by default)\n");
	printf(" -P        : generate prefix-list (default, just for backward"
		" compatibility)\n");
	printf(" -R len    : allow more specific routes up to specified masklen\n");
	printf(" -r len    : allow more specific routes from masklen specified\n");
	printf(" -S sources: use only specified sources (recommended:"
		" RADB,RIPE,APNIC)\n");
	printf(" -s        : generate sequence numbers in prefix-lists (IOS only)\n");
	printf(" -T        : disable pipelining (experimental, faster mode)\n");
	printf(" -t        : generate as-sets for OpenBGPD (OpenBSD 6.4+), BIRD "
		"and JSON\n");
	printf("             formats, as-list-groups for Juniper (JunOS 21+)\n");
	printf(" -U        : generate config for Huawei (Cisco IOS by default)\n");
	printf(" -W len    : specify max-entries on as-path line (use 0 for "
		"infinity)\n");
	printf(" -w        : 'validate' AS numbers: accept only ones with "
		"registered routes\n");
	printf(" -X        : generate config for IOS XR (Cisco IOS by default)\n");
	printf(" -x        : generate mixed-family (both IPv4 and IPv6) prefix "
		"filters\n");
	printf(" -z        : generate route-filter-list (Juniper only)\n");
	printf("\n" PACKAGE_NAME " version: " PACKAGE_VERSION "\n");
	printf("Copyright(c) Alexandre Snarskii <snar@snar.spb.ru> 2007-2026\n\n");
	exit(ecode);
};

void
exclusive()
{
	fprintf(stderr,"-E, -f <asnum>, -G <asnum>, -P, -t and -z are mutually "
		"exclusive\n");
	exit(1);
};

void
vendor_exclusive()
{
	fprintf(stderr, "-b (BIRD), -B (OpenBGPD), -F (formatted), -J (JunOS), "
		"-j (JSON), -N (Nokia SR OS classic), -n (Nokia SR OS MD-CLI), "
		"-U (Huawei) and -X (IOS XR) options are mutually exclusive\n");
	exit(1);
};

int
parseasnumber(struct bgpq_expander* expander, char* optarg, int zero)
{
	char* eon=NULL;
	expander->asnumber=strtoul(optarg,&eon,10);
	if((!zero && expander->asnumber<1) ||
		expander->asnumber>(65535ul*65535)) {
		sx_report(SX_FATAL,"Invalid AS number: %s\n", optarg);
		exit(1);
	};
	if(eon && *eon=='.') {
		/* -f 3.3, for example */
		uint32_t loas=strtoul(eon+1,&eon,10);
		if(expander->asnumber>65535) {
			/* should prevent incorrect numbers like 65537.1 */
			sx_report(SX_FATAL,"Invalid AS number: %s\n", optarg);
			exit(1);
		};
		if(loas<1 || loas>65535) {
			sx_report(SX_FATAL,"Invalid AS number: %s\n", optarg);
			exit(1);
		};
		if(eon && *eon) {
			sx_report(SX_FATAL,"Invalid symbol in AS number: %c (%s)\n",
				*eon, optarg);
			exit(1);
		};
		expander->asnumber=(expander->asnumber<<16)+loas;
	} else if(eon && *eon) {
		sx_report(SX_FATAL,"Invalid symbol in AS number: %c (%s)\n",
			*eon, optarg);
		exit(1);
	};
	return 0;
};

int
main(int argc, char* argv[])
{
	int c;
	struct bgpq_expander expander;
	int af=AF_INET, selectedipv4 = 0, exceptmode = 0;
	int widthSet=0, aggregate=0, refine=0, refineLow=0, hyperaggregate=0;
	unsigned long maxlen=0;

	bgpq_expander_init(&expander,af);
	if (getenv("IRRD_SOURCES"))
		expander.sources=getenv("IRRD_SOURCES");

	while((c=getopt(argc,argv,"2346a:AbBdDEF:HS:jJf:l:L:m:M:NnW:Ppr:R:G:tTh:I:UwXxsz"))
		!=EOF) {
	switch(c) {
		case '2':
			expand_as23456=1;
			break;
		case '3':
			expander.asn32=1;
			break;
		case '4':
			/* do nothing, expander already configured for IPv4 */
			if (expander.family == AF_INET6 || expander.treex) {
				sx_report(SX_FATAL, "-4, -6 and -x are mutually exclusive\n");
				exit(1);
			};
			selectedipv4=1;
			break;
		case '6':
			if (selectedipv4 || expander.treex) {
				sx_report(SX_FATAL, "-4, -6 and -x are mutually exclusive\n");
				exit(1);
			};
			af=AF_INET6;
			expander.family=AF_INET6;
			expander.tree->family=AF_INET6;
			break;
		case 'a':
			parseasnumber(&expander,optarg,0);
			break;
		case 'A':
			if(hyperaggregate) {
				sx_report(SX_FATAL, "-A and -H are mutually exclusive\n");
				exit(1);
			};
			if(aggregate) debug_aggregation++;
			aggregate=1;
			break;
		case 'b':
			if(expander.vendor) vendor_exclusive();
			expander.vendor=V_BIRD;
			break;
		case 'B':
			if(expander.vendor) vendor_exclusive();
			expander.vendor=V_OPENBGPD;
			expander.asn32=1;
			break;
		case 'D': expander.asdot=1;
			break;
		case 'd': debug_expander++;
			break;
		case 'E': if(expander.generation) exclusive();
			expander.generation=T_EACL;
			break;
		case 'F': if(expander.vendor) exclusive();
			expander.vendor=V_FORMAT;
			expander.format=optarg;
			break;
		case 'f':
			if(expander.generation) exclusive();
			expander.generation=T_ASPATH;
			parseasnumber(&expander,optarg,1);
			break;
		case 'G':
			if(expander.generation) exclusive();
			expander.generation=T_OASPATH;
			parseasnumber(&expander,optarg,0);
			break;
		case 'H':
			if(aggregate) {
				sx_report(SX_FATAL, "-A and -H are mutually exclusive\n");
				exit(1);
			};
			hyperaggregate=1;
			break;
		case 'h': {
			char* d=strchr(optarg, ':');
			expander.server=optarg;
			if(d) {
				*d=0;
				expander.port=d+1;
			};
			break;
		};
		case 'I':
			expander.laddr=optarg;
			break;
		case 'J': if(expander.vendor) vendor_exclusive();
			expander.vendor=V_JUNIPER;
			break;
		case 'j': if(expander.vendor) vendor_exclusive();
			expander.vendor=V_JSON;
			break;
		case 'p':
			expand_special_asn=1;
			break;
		case 'P':
			if(expander.generation) exclusive();
			expander.generation=T_PREFIXLIST;
			break;
		case 'r':
			refineLow=strtoul(optarg,NULL,10);
			if(!refineLow) {
				sx_report(SX_FATAL,"Invalid refineLow value: %s\n", optarg);
				exit(1);
			};
			break;
		case 'R':
			refine=strtoul(optarg,NULL,10);
			if(!refine) {
				sx_report(SX_FATAL,"Invalid refine length: %s\n", optarg);
				exit(1);
			};
			break;
		case 'l': expander.name=optarg;
			break;
		case 'L': expander.maxdepth=strtol(optarg, NULL, 10);
			if (expander.maxdepth < 1) {
				sx_report(SX_FATAL, "Invalid maximum recursion (-L): %s\n",
					optarg);
				exit(1);
			};
			break;
		case 'm': maxlen=strtoul(optarg, NULL, 10);
			if (!maxlen) {
				sx_report(SX_FATAL, "Invalid maxlen (-m): %s\n", optarg);
				exit(1);
			};
			break;
		case 'M': {
			char* c, *d;
			expander.match=strdup(optarg);
			c=d=expander.match;
			while(*c) {
				if(*c=='\\') {
					if(*(c+1)=='n') {
						*d='\n';
						d++;
						c+=2;
					} else if(*(c+1)=='r') {
						*d='\r';
						d++;
						c+=2;
					} else if(*(c+1)=='t') {
						*d='\t';
						d++;
						c+=2;
					} else if(*(c+1)=='\\') {
						*d='\\';
						d++;
						c+=2;
					} else {
						sx_report(SX_FATAL, "Unsupported escape \%c (0x%2.2x) "
							"in '%s'\n", isprint(*c)?*c:20, *c, optarg);
						exit(1);
					};
				} else {
					if(c!=d) {
						*d=*c;
					};
					d++;
					c++;
				};
			};
			*d=0;
			};
			break;
		case 'N': if(expander.vendor) vendor_exclusive();
			expander.vendor=V_NOKIA;
			break;
		case 'n': if(expander.vendor) vendor_exclusive();
			expander.vendor=V_NOKIA_MD;
			break;
		case 't':
			if(expander.generation) exclusive();
			expander.generation=T_ASSET;
			break;
		case 'T': pipelining=0;
			break;
		case 's': expander.sequence=1;
			break;
		case 'S': expander.sources=optarg;
			break;
		case 'U':
			if(expander.vendor) vendor_exclusive();
			expander.vendor=V_HUAWEI;
			break;
		case 'W': expander.aswidth=atoi(optarg);
			if(expander.aswidth<0) {
				sx_report(SX_FATAL,"Invalid as-width: %s\n", optarg);
				exit(1);
			};
			widthSet=1;
			break;
		case 'w': expander.validate_asns=1;
			break;
		case 'X': if(expander.vendor) vendor_exclusive();
			expander.vendor=V_CISCO_XR;
			break;
		case 'x':
			if (selectedipv4 || expander.family==AF_INET6) {
				sx_report(SX_FATAL, "-4, -6 and -x are mutually exclusive\n");
				exit(1);
			};
			expander.treex = sx_radix_tree_new(AF_INET6);
			if (!expander.treex) {
				sx_report(SX_FATAL, "error initializing treex: %s\n",
					strerror(errno));
				exit(1);
			};
			break;
		case 'z':
			if(expander.generation) exclusive();
			expander.generation=T_ROUTE_FILTER_LIST;
			break;
		default : usage(1);
	};
	};

	argc-=optind;
	argv+=optind;

	if(!widthSet) {
		if(expander.generation==T_ASPATH) {
			if(expander.vendor==V_CISCO) {
				expander.aswidth=4;
			} else if(expander.vendor==V_CISCO_XR) {
				expander.aswidth=6;
			} else if(expander.vendor==V_JUNIPER) {
				expander.aswidth=8;
			} else if(expander.vendor==V_BIRD) {
				expander.aswidth=10;
			} else if(expander.vendor==V_NOKIA || expander.vendor==V_NOKIA_MD) {
				expander.aswidth=8;
			};
		} else if(expander.generation==T_OASPATH) {
			if(expander.vendor==V_CISCO) {
				expander.aswidth=5;
			} else if(expander.vendor==V_CISCO_XR) {
				expander.aswidth=7;
			} else if(expander.vendor==V_JUNIPER) {
				expander.aswidth=8;
			} else if(expander.vendor==V_NOKIA || expander.vendor==V_NOKIA_MD) {
				expander.aswidth=8;
			};
		} else if(expander.generation==T_ASSET) {
			if(expander.vendor==V_JUNIPER) {
				expander.aswidth=7;
			};
		};
	};

	if(!expander.generation) {
		expander.generation=T_PREFIXLIST;
	};

	if(expander.vendor==V_CISCO_XR && expander.generation!=T_PREFIXLIST &&
		expander.generation!=T_ASPATH && expander.generation!=T_OASPATH) {
		sx_report(SX_FATAL, "Sorry, only prefix-sets and as-paths "
			"supported for IOS XR\n");
	};
	if(expander.vendor==V_BIRD && expander.generation!=T_PREFIXLIST &&
		expander.generation!=T_ASPATH && expander.generation!=T_ASSET) {
		sx_report(SX_FATAL, "Sorry, only prefix-lists and as-paths/as-sets "
			"supported for BIRD output\n");
	};
	if(expander.vendor==V_JSON && expander.generation!=T_PREFIXLIST &&
		expander.generation!=T_ASPATH && expander.generation!=T_ASSET) {
		sx_report(SX_FATAL, "Sorry, only prefix-lists and as-paths/as-sets "
			"supported for JSON output\n");
	};
	if(expander.vendor==V_FORMAT && expander.generation!=T_PREFIXLIST)
		sx_report(SX_FATAL, "Sorry, only prefix-lists supported in formatted "
			"output\n");
	if(expander.vendor==V_FORMAT && (refine || refineLow)) {
		sx_report(SX_FATAL, "Sorry, formatted output (-F <fmt>) in not "
			"compatible with -R/-r options\n");
		exit(1);
	};
	if(expander.vendor==V_HUAWEI && expander.generation!=T_ASPATH &&
		expander.generation!=T_OASPATH && expander.generation != T_PREFIXLIST)
		sx_report(SX_FATAL, "Sorry, only as-paths and prefix-lists supported "
			"for Huawei output\n");

	if(expander.generation==T_ROUTE_FILTER_LIST && expander.vendor!=V_JUNIPER) {
		sx_report(SX_FATAL, "Route-filter-lists (-z) supported for Juniper (-J)"
			" output only\n");
	};
	if(expander.generation==T_ASSET && expander.vendor!=V_JSON &&
		expander.vendor!=V_OPENBGPD && expander.vendor!=V_BIRD &&
		expander.vendor!=V_JUNIPER) {
		sx_report(SX_FATAL, "As-Sets/As-Lists (-t) supported for JSON (-j), "
			"OpenBGPD (-B), BIRD (-b) and Juniper (-J) output only\n");
	};

	if(expander.asdot && expander.vendor!=V_CISCO) {
		sx_report(SX_FATAL,"asdot notation supported only for Cisco, "
			"other formats use asplain only\n");
	};

	if(!expander.asn32 && expander.asnumber>65535) {
		expander.asnumber=23456;
	};

	if(aggregate && expander.vendor==V_JUNIPER &&
		expander.generation==T_PREFIXLIST) {
		sx_report(SX_FATAL, "Sorry, aggregation (-A) does not work in"
			" Juniper prefix-lists\nYou can try route-filters (-E) "
			"or route-filter-lists (-z) instead of prefix-lists "
			"(-P, default)\n");
		exit(1);
	};

	if(aggregate && expander.vendor==V_FORMAT) {
		sx_report(SX_FATAL, "Sorry, aggregation (-A) is not compatible with "
			"formatted output (-F <fmt>)\n");
		exit(1);
	};

	if(aggregate && (expander.vendor==V_NOKIA_MD || expander.vendor==V_NOKIA) &&
		expander.generation!=T_PREFIXLIST) {
		sx_report(SX_FATAL, "Sorry, aggregation (-A) is not supported with "
			"ip-prefix-lists (-E) on Nokia. You can try prefix-lists (-P) "
			"instead\n");
		exit(1);
	};
	if(refine && (expander.vendor==V_NOKIA_MD || expander.vendor==V_NOKIA) &&
		expander.generation!=T_PREFIXLIST) {
		sx_report(SX_FATAL, "Sorry, more-specifics (-R) is not supported with "
			"ip-prefix-lists (-E) on Nokia. You can try prefix-lists (-P) "
			"instead\n");
		exit(1);
	};
	if(refineLow && (expander.vendor==V_NOKIA_MD || expander.vendor==V_NOKIA) &&
		expander.generation!=T_PREFIXLIST) {
		sx_report(SX_FATAL, "Sorry, more-specifics (-r) is not supported with "
			"ip-prefix-lists (-E) on Nokia. You can try prefix-lists (-P) "
			"instead\n");
		exit(1);
	};

	if(aggregate && expander.generation<T_PREFIXLIST) {
		sx_report(SX_FATAL, "Sorry, aggregation (-A) used only for prefix-"
			"lists, extended access-lists and route-filters\n");
		exit(1);
	};

	if(hyperaggregate && expander.generation<T_PREFIXLIST) {
		sx_report(SX_FATAL, "Sorry, hyperaggregation (-H) used only for "
			"prefix-lists, extended access-lists and route-filters\n");
		exit(1);
	};

	if(hyperaggregate && (refine || refineLow)) {
		sx_report(SX_FATAL, "Sorry, more-specifics (-R/-r) make no sense "
			"in hyperaggregation/supernets-only (-H) mode\n");
		exit(1);
	};
	if (expander.treex && (refine || refineLow)) {
		sx_report(SX_FATAL, "Sorry, more-specifics (-R/-r) make no sense "
			"in mixed-af (-x) mode\n");
		exit(1);
	};
	if (expander.treex && expander.generation<T_PREFIXLIST) {
		sx_report(SX_FATAL, "Sorry, mixed-af (-x) mode can't be used for "
			"non prefix-lists\n");
		exit(1);
	};
	if (expander.treex && maxlen !=0) {
		sx_report(SX_FATAL, "Sorry, max prefix length filter (-m) can't "
			"be used with mixed-af (-x) mode\n");
		exit(1);
	};
	if (expander.treex && !(expander.vendor == V_JUNIPER ||
		expander.vendor == V_JSON || expander.vendor == V_FORMAT)) {
		sx_report(SX_FATAL, "Sorry, mixed-af (-x) mode implemened for "
			"Juniper (-J), JSON (-j) or User-Defined (-F) formats only\n");
		exit(1);
	};

	if (expander.sequence && expander.vendor!=V_CISCO) {
		sx_report(SX_FATAL, "Sorry, prefix-lists sequencing (-s) supported"
			" only for IOS\n");
		exit(1);
	};

	if (expander.sequence && expander.generation<T_PREFIXLIST) {
		sx_report(SX_FATAL, "Sorry, prefix-lists sequencing (-s) can't be "
			" used for non prefix-list\n");
		exit(1);
	};

	if(refineLow && !refine) {
		if(expander.family == AF_INET)
			refine = 32;
		else
			refine = 128;
	};

	if (refineLow && refineLow > refine) {
		sx_report(SX_FATAL, "Incompatible values for -r %u and -R %u\n",
			refineLow, refine);
	};

	if(refine || refineLow) {
		if(expander.family==AF_INET6 && refine>128) {
			sx_report(SX_FATAL, "Invalid value for refine(-R): %u (1-128 for"
				" IPv6)\n", refine);
		} else if(expander.family==AF_INET6 && refineLow>128) {
			sx_report(SX_FATAL, "Invalid value for refineLow(-r): %u (1-128 for"
				" IPv6)\n", refineLow);
		} else if(expander.family==AF_INET && refine>32) {
			sx_report(SX_FATAL, "Invalid value for refine(-R): %u (1-32 for"
				" IPv4)\n", refine);
		} else if(expander.family==AF_INET && refineLow>32) {
			sx_report(SX_FATAL, "Invalid value for refineLow(-r): %u (1-32 for"
				" IPv4)\n", refineLow);
		};

		if(expander.vendor==V_JUNIPER && expander.generation==T_PREFIXLIST) {
			if(refine) {
				sx_report(SX_FATAL, "Sorry, more-specific filters (-R %u) "
					"is not supported for Juniper prefix-lists.\n"
					"Use router-filters (-E) or route-filter-lists (-z) "
					"instead\n", refine);
			} else {
				sx_report(SX_FATAL, "Sorry, more-specific filters (-r %u) "
					"is not supported for Juniper prefix-lists.\n"
					"Use route-filters (-E) or route-filter-lists (-z) "
					"instead\n", refineLow);
			};
		};

		if(expander.generation<T_PREFIXLIST) {
			if(refine) {
				sx_report(SX_FATAL, "Sorry, more-specific filter (-R %u) "
		 			"supported only with prefix-list generation\n", refine);
			} else {
				sx_report(SX_FATAL, "Sorry, more-specific filter (-r %u) "
					"supported only with prefix-list generation\n", refineLow);
			};
		};
	};

	if(maxlen) {
		if((expander.family==AF_INET6 && maxlen>128) ||
			(expander.family==AF_INET  && maxlen>32)) {
			sx_report(SX_FATAL, "Invalid value for max-prefixlen: %lu (1-128 "
				"for IPv6, 1-32 for IPv4)\n", maxlen);
			exit(1);
		} else if((expander.family==AF_INET6 && maxlen<128) ||
			(expander.family==AF_INET  && maxlen<32)) {
			/* inet6/128 and inet4/32 does not make sense - all routes will
			 * be accepted, so save some CPU cycles :) */
			expander.maxlen = maxlen;
		};
	} else if (expander.family==AF_INET) {
		expander.maxlen = 32;
	} else if (expander.family==AF_INET6) {
		expander.maxlen = 128;
	};

	if(expander.generation==T_EACL && expander.vendor==V_CISCO &&
		expander.family==AF_INET6) {
		sx_report(SX_FATAL,"Sorry, ipv6 access-lists not supported for Cisco"
			" yet.\n");
	};

	if(expander.match != NULL && (expander.vendor != V_JUNIPER ||
		expander.generation != T_EACL)) {
		sx_report(SX_FATAL, "Sorry, extra match conditions (-M) can be used "
			"only with Juniper route-filters\n");
	};

	if((expander.generation==T_ASPATH || expander.generation==T_OASPATH) &&
		af != AF_INET && !expander.validate_asns) {
		sx_report(SX_FATAL, "Sorry, -6 makes no sense with as-path (-f/-G) "
			"generation\n");
	};

	if (expander.validate_asns && expander.generation != T_ASPATH &&
		expander.generation != T_OASPATH && expander.generation != T_ASSET) {
		sx_report(SX_FATAL, "Sorry, -w makes sense only for as-path (-f/-G) "
			"or as-set/as-list (-t) generation\n");
	};
	if(expander.generation==T_ASPATH && expander.asnumber==0 &&
		(expander.vendor==V_OPENBGPD)) {
		sx_report(SX_FATAL, "Sorry, -f 0 makes no sense with OpenBGPD\n");
	};

	if(!argv[0])
		usage(1);

	while(argv[0]) {
		if(!strcmp(argv[0], "EXCEPT")) {
			exceptmode = 1;
		} else if (exceptmode) {
			bgpq_expander_add_stop(&expander,argv[0]);
		} else if(!strncasecmp(argv[0],"AS-",3)) {
			bgpq_expander_add_asset(&expander,argv[0]);
		} else if(!strncasecmp(argv[0],"RS-",3)) {
			bgpq_expander_add_rset(&expander,argv[0]);
		} else if(!strncasecmp(argv[0],"AS",2)) {
			char* c;
			if((c=strchr(argv[0],':'))) {
				if(!strncasecmp(c+1,"AS-",3)) {
					bgpq_expander_add_asset(&expander,argv[0]);
				} else if(!strncasecmp(c+1,"RS-",3)) {
					bgpq_expander_add_rset(&expander,argv[0]);
				} else {
					SX_DEBUG(debug_expander,"Unknown sub-as object %s\n",
						argv[0]);
				};
			} else {
				bgpq_expander_add_as(&expander,argv[0]);
			};
		} else {
			char* c = strchr(argv[0], '^');
			if (!c && !bgpq_expander_add_prefix(&expander,argv[0])) {
				sx_report(SX_ERROR, "Unable to add prefix %s (bad prefix or "
					"address-family)\n", argv[0]);
				exit(1);
			} else if (c && !bgpq_expander_add_prefix_range(&expander,argv[0])){
				sx_report(SX_ERROR, "Unable to add prefix-range %s (bad range "
					"or address-family)\n", argv[0]);
				exit(1);
			};
		};
		argv++;
		argc--;
	};

	if(!bgpq_expand(&expander)) {
		exit(1);
	};

	if(refine)
		sx_radix_tree_refine(expander.tree,refine);

	if(refineLow)
		sx_radix_tree_refineLow(expander.tree, refineLow);

	if(aggregate || hyperaggregate) {
		sx_radix_tree_aggregate(expander.tree);
		if (expander.treex)
			sx_radix_tree_aggregate(expander.treex);
	};

	if(hyperaggregate) {
		sx_radix_tree_hyperaggregate(expander.tree);
		if (expander.treex)
			sx_radix_tree_hyperaggregate(expander.treex);
	};

	switch(expander.generation) {
		case T_NONE: sx_report(SX_FATAL,"Unreachable point... call snar\n");
			exit(1);
		case T_ASPATH: bgpq3_print_aspath(stdout,&expander);
			break;
		case T_OASPATH: bgpq3_print_oaspath(stdout,&expander);
			break;
		case T_ASSET: bgpq3_print_asset(stdout,&expander);
			break;
		case T_PREFIXLIST: bgpq3_print_prefixlist(stdout,&expander);
			break;
		case T_EACL: bgpq3_print_eacl(stdout,&expander);
			break;
		case T_ROUTE_FILTER_LIST:
			bgpq3_print_route_filter_list(stdout, &expander);
			break;
	};

	return 0;
};

